table/tables/index.go

```diff
@@ -400,6 +410,23 @@ func (c *index) Delete(sc *stmtctx.StatementContext, txn kv.Transaction, indexed
 		}
 		tempValElem := tablecodec.TempIndexValueElem{Handle: h, KeyVer: tempKeyVer, Delete: true, Distinct: distinct}
 
+		// If index is global, decode the pid from value (if exists) and compare with c.physicalID.
+		// Only when pid in value equals to c.physicalID, the key can be deleted.
+		if c.idxInfo.Global {
+			if val, err := txn.GetMemBuffer().Get(context.Background(), key); err == nil {
+				segs := tablecodec.SplitIndexValue(val)
+				if len(segs.PartitionID) != 0 {
+					_, pid, err := codec.DecodeInt(segs.PartitionID)
+					if err != nil {
+						return err
+					}
+					if pid != c.phyTblID {
+						continue
+					}
+				}
+			}
+		}
```

Comment:
I was a bit curious and I wonder if we should change the way we add the indexes [here](https://github.com/pingcap/tidb/blob/6897c1abb8adb782a474ceaeda005f6cddc10013/table/tables/tables.go#L186-L198)?
I mean if it is a global index, why add it with the physical table ids of the partitions?